Oracle checking using SNYK data

shikhar394 opened the pull request:
Greetings, 
The following pull request completes the task of version checking against the oracle. 
- Added a new file in the vulnerabilities folder. VulnerabilityMap.json (see 
ea3690b) that holds the relevant information from the SNYK data. The map is created by the file generate_SNYK_vulnerabilities.py (see ea3690b).

- Created a new class to help with the comparison and handling of the ranges. Included tests. (see 
d77b95b)

- Made changes to the mongo_handler.py to accommodate changes in the vulnerabilities folder and to use the version-comparison class to abstract the comparison. (see a4d8df1)

- Added semantic_version to requirements.txt (see 00f80b5)

TODO (Tentative):
Scrape GitHub packages to check for tags.

shikhar394 committed:
- Code to convert SNYK data from maven.json (not committed) to a json file. JSON Schema: CVEXXX:{PackageName:XXX, Version:[XXX]}
- VulnerabilityMap.json contains the json created from the code.

shikhar394 committed:
Created class that takes the version range from the VulnerabilityMap.json and an arbitrary range and compares them. Using the semantic_version Python package.

shikhar394 committed:
Added semantic_version

shikhar394 committed:
Changed code to accomodate oracle checking using the VulnerabilityMap.json and VersionComparisonClass

shikhar394 committed:
Added check for undefined upper limit

cassianomonteiro commented on VersionComparisonClass.py:
```diff
@@ -0,0 +1,91 @@
+    VersionRangeList = []
+    for versionPoint in self.VersionRangeInput:
+      VersionRangeList.append(self.UnpackRange(versionPoint))
+      if len(VersionRangeList) == 2:
```
Why do you check for length 2? What if there is only 1 version affected by a vulnerability?

shikhar394 replied:
Thanks for pointing that out. Check 47b6307.

cassianomonteiro reviewed (changes requested):
Don't commit VulnerabilityMap.json, that can be generated whenever we have new data from SNYK.

cassianomonteiro commented on VersionComparisonClass.py:
```diff
@@ -0,0 +1,91 @@
+  """
+  Test
+  """
+  VersionRange = ['[3.2-alpha,3.3-beta-2)']
```
Does it work with versions like 'jenkins-1.2.3.4' (with a prefix)?

shikhar394 replied:
Thanks for pointing that out. 
I made a fix for that see 374fa11, ce57ecb. 
This is more like a bandaid fix. Should work in theory. Also working in the test cases. Needs more testing to check the efficacy of the fix.
It converts the prefix into a suffix which then gets treated as a regular patch name, so think: Jenkins < Prototype. I've tested this with some variations like Jenkins-1.2.3-alpha which works. 
What do you think?

Of course, the good way to fix this would be to comb through the github tags and work with them. But even those would have to be worked in a similar manner, I think.

cassianomonteiro commented on mongo_handler.py:
```diff
@@ -169,24 +174,23 @@ def dump_vulnerabilities(collection):
-vulnerabilities_map = {}
-
-
-def get_oracle(vulnerability_id, version):
```
Please keep the old code for the vulnerabilities already mapped. If the file '..._vulnerability.txt'  is not found, than do your check.

shikhar394 replied:
There you go! 9af1229

cassianomonteiro commented on VersionComparisonClass.py:
```diff
@@ -0,0 +1,91 @@
+    elif versionPoint[-1] == ')':
+      return '<' + str(sv.Version.coerce(versionPoint[:-1]))
+    else:
+      raise Exception("Range not recongized")
```
In this case, let's not raise an Exception. We need to label this cases for manual verification.

shikhar394 replied:
Got it! See, a4f186f
Thanks!

cassianomonteiro commented on generate_SNYK_vulnerabilities.py:
```diff
@@ -0,0 +1,23 @@
+from pprint import pprint
+
+def BuildVulnerabilityMap():
+  CVE_Version = {} #CVE:{package_name:XXX, version:XXX}
```
Some vulnerabilities don't have a CVE ID. Let's use SNYK ID.

shikhar394 replied:
Do you mean only for the vulnerabilities that don't have a CVE ID or all of them? 

I don't think using them for all of them would be helpful because the search would be very expensive. Currently, since the mongo_handler passes in a vulnerability ID, I indexed the dictionary using the same. 

Can we change the mongo_handler (and the algorithm) to return an ID other than the CVE ID, if you do suggest changing all the IDs to SNYK ID?

Thanks!

cassianomonteiro replied:
Yes, let's use the SNYK ID for all of them. I will change mongo_handler to do the same.

shikhar394 replied:
There you go! 93c791d

shikhar394 committed:
Added fix for handling single vulnerability

shikhar394 committed:
Addresses version prefixes.
- Almost a bandaid fix. Adressesses only prefixes that are entirely strings (Jenkins, Prototype, etc.
- Converts the prefix into a suffix, this is easily handled by the package.
- Need to test with wild data to see how effective the solution is. In theory, it should work.
- TODO: Github Tags

shikhar394 committed:
Fixed some comments from previous commit

shikhar394 committed:
Readded original code for checking vulnerabilities. If that returns False, then checks snyk data.

shikhar394 committed:
Removed exception for unrecognized version format. Writes to 'experiments/vulnerabilities/ManualInspectionLog.txt'

cassianomonteiro commented on VersionComparisonClass.py:
```diff
@@ -0,0 +1,183 @@
+    ['Jenkins-1.2.3' or 'Jenkins-1.2.3]
+    Checks and recreates the version properly.
+    '''
+    VersionParts = versionPoint.split('-')
```
This might not be the best idea, as other tags might have different ways to include prefixes (such with other characters like underscore, dot, etc.). Can you try using some regex to catch the #.#.#... format?

shikhar394 replied:
Great suggestion. Here you go 101c713.

shikhar394 committed:
Added regex version selection in versions with prefix

shikhar394 committed:
Indexing the vulnerability map with SNYK ID

shikhar394 committed:
removed file
